File: cases/taint/java/dispatch-table-negative/DispatchTableNegative.java
package dataflowbench.taint;

import java.util.HashMap;
import java.util.Map;
import java.util.function.IntUnaryOperator;

final class DispatchTableNegative {
    static int dfb_source() { // DFB-SOURCE: dispatch-table-input
        return 1;
    }

    static void dfb_sink(int value) { } // DFB-SINK: dispatch-table-sink

    static void run() {
        Map<String, IntUnaryOperator> table = new HashMap<>();
        table.put("leak", value -> {
            dfb_sink(value);
            return value;
        });
        table.put("drop", value -> { // DFB-WITNESS: dispatch-table-entry
            dfb_sink(0);
            return 0;
        });
        String key = "drop";
        table.get(key).applyAsInt(dfb_source()); // DFB-WITNESS: dispatch-table-lookup
    }
}

File: cases/taint/rust/dispatch-table-negative/dispatch_table_negative.rs
use std::collections::HashMap;

fn dfb_source() -> i32 { // DFB-SOURCE: dispatch-table-input
    1
}

fn dfb_sink(value: i32) {} // DFB-SINK: dispatch-table-sink

fn leak(value: i32) { // DFB-WITNESS: dispatch-table-entry
    dfb_sink(value);
}

fn drop_argument(value: i32) {
    dfb_sink(0);
}

fn run() {
    let mut table: HashMap<&str, fn(i32)> = HashMap::new();
    table.insert("leak", leak);
    table.insert("drop", drop_argument);
    let key = "drop";
    let selected = table[key];
    selected(dfb_source());
}
